feat: allow spectral rules to be defined in the validaterc

rmkeezer opened the pull request:
ref: https://github.ibm.com/arf/planning-sdk-squad/issues/2170
followup of: https://github.com/IBM/openapi-validator/pull/195

These changes allow spectral rules to have their severity level changed in the `.validatrc` and allow custom rules defined in a `.spectral.yaml` file to not be overridden by the `.defaultsForSpectral.yaml`. Although the `.defaultsForSpectral.yaml` will still override the `.spectral.yaml` severity level of any existing rule `.defaultsForSpectral.yaml` defines a severity for, including the extended rules.

rmkeezer commented on src/cli-validator/utils/processConfiguration.js:
```diff
@@ -42,102 +42,112 @@ const validateConfigObject = function(configObject, chalk) {
-    userCategories.forEach(category => {
-      if (!allowedCategories.includes(category)) {
+    // Do not check "spectral" spec rules
+    if (spec !== 'spectral') {
```
This entire file is just indention changes with a single if statement, not sure why the github diff shows it so poorly

jorge-ibm commented on src/spectral/rulesets/.defaultsForSpectral.yaml:
```diff
@@ -1,6 +1,6 @@
 extends: [[spectral:oas, off]]
 formats: [oas2, oas3]
-functionsDir": ../functions
+functionsDir: ../functions
```
Surprised spectral didn't complain about that

jorge-ibm commented on src/spectral/utils/spectral-validator.js:
```diff
@@ -57,12 +62,27 @@ const setup = async function(spectral) {
+  // Combine user ruleset with the default ruleset
+  // The default ruleset will take precendence in any rules it defines (including extended rules)
+  const defaultRuleset = await readRuleset(defaultSpectralRulesetURI);
+  mergeRules(spectralRuleset.rules, defaultRuleset.rules);
```
Might be worth trying to see if you can achieve the same merging by simply calling ` await spectral.loadRuleset([spectralRuleset, defaultSpectralRulesetURI]);`

rmkeezer replied:
Yup this does achieve the same, I wasn't confident it did at first but it seems to be working as expect! I updated the PR with this

jorge-ibm reviewed (changes requested):
It's looking better, just a couple of more things that need to be changed. Also, if possible, some unit tests to test the validaterc changes would be helpful.

rmkeezer commented on src/.defaultsForValidator.js:
```diff
@@ -106,6 +106,30 @@ const defaults = {
       'json_or_param_binary_string': 'warning'
     }
+  },
+  'spectral': {
```
Since i'm calling `getConfigObject` if there is no spectral section in the validaterc, it will take the rules from the spectral rules from `.defaultsForValidator.js` and these changes also allow the `init` command to generate the new spectral section

rmkeezer commented on src/spectral/utils/spectral-validator.js:
```diff
@@ -57,12 +57,19 @@ const setup = async function(spectral) {
 
+  // Combine user ruleset with the default ruleset
+  // The default ruleset will take precendence in any rules it defines (including extended rules)
+  await spectral.loadRuleset([spectralRulesetURI, defaultSpectralRulesetURI]);
```
I dont think there is a problem with loading a ruleset twice since it won't create any duplicates. Currently it will load the user config and the `.defaultsForSpectral.yaml` config or the `.defaultsForSpectral.yaml` config twice, how should this be changed? I can change it to load the `.defaultsForSpectral.yaml` first, should I make any other changes?

jorge-ibm replied:
Okay. Yes, I think we should load the ``.defaultsForSpectral.yaml `` first

rmkeezer replied:
Ok i updated it to load the defaults first

jorge-ibm commented:
@rmkeezer I'm not sure if my suggestion to use ``loadRuleset`` vs ``mergeRules`` broke something, but I pulled your fork locally and wasn't able to overwrite the severity level of our rules using the validaterc file.

rmkeezer commented:
@jorge-ibm Ah, apparently `spectral.mergeRules` does not work but using the `mergeRules` function from the spectral ruleset does. I've updated the PR with that change, it should now correctly override the severity again.

jorge-ibm commented:
Looks like everything is working as expected now, though I still think some tests would be helpful. The only thing I think should be fixed before merging is this duplicate message displaying whenever the validaterc file isn't present:
```
JorgeWoComputer:openapi-validator-rkeezer jrangel$ lint-openapi -p test/spectral/mockFiles/oas3/enabled-rules.yml 
[Warning] No .validaterc file found. The validator will run in default mode.
To configure the validator, create a .validaterc file.
[Warning] No .validaterc file found. The validator will run in default mode.
To configure the validator, create a .validaterc file.
```

I'm assuming we want to add documentation on how to enable/edit our spectral rules from the validaterc file. @dpopp07 what are your thoughts?

jorge-ibm reviewed (changes requested):
Sorry for requesting another set of changes, but I think this PR is really close.

jorge-ibm reviewed (approved):
LGTM. I'll approve but we should hold off on merging until @dpopp07 or @mkistler review it as well

dpopp07 commented on src/spectral/utils/spectral-validator.js:
```diff
@@ -46,7 +48,7 @@ const parseResults = function(results, debug) {
 
 // setup: registers the oas2/oas3 formats, and attempts to load the ruleset file
-const setup = async function(spectral) {
+const setup = async function(spectral, defaultMode = false) {
```
Is there anywhere `setup` would be called without specifying this value? It looks like the one place we call it, we provide the value. So this default seems redundant. If it can be called in other places, that's a different story!

dpopp07 replied:
Okay, I found it 🙂 This happens [on this line](https://github.com/IBM/openapi-validator/blob/master/src/lib/index.js#L18) but we have the default value there as a variable - so I think it would be better to pass it in that let it be potentially redefined. This should be addressed before approval

dpopp07 commented on src/spectral/utils/spectral-validator.js:
```diff
@@ -57,12 +59,22 @@ const setup = async function(spectral) {
 
+  // Combine user ruleset with the default ruleset
+  // The default ruleset will take precendence in any rules it defines (including extended rules)
+  await spectral.loadRuleset([defaultSpectralRulesetURI, spectralRulesetURI]);
```
So this handles the case in which users define their own configuration with the Spectral format? But we don't let them override our default Spectral rules with that format?

I think this is what we want but as someone not very familiar with Spectral, I find it a little confusing. I appreciate the comments here, they are helpful, but will you add a couple more to provide context for why this is there?

jorge-ibm replied:
Matthew correct me if I'm wrong but with this PR this is what the user can do now:
1. They can override the severity level and disable/enable our spectral rules from the validaterc file.
2. They can _add_ any new custom rules to the validator using a custom rule set. Even if they don't include our spectral rules in their ruleset file, our spectral rules will still be loaded.
3. If a user wants to override our spectral rules using their ruleset file, they would have to:
   - remove that spectral rule from the validaterc file
   - add a rule with the same name to their ruleset file

rmkeezer replied:
1. Correct
2. Correct
3. Correct, but I realized that the omitting the spectral rule from the validaterc file did _not_ remove it and still overrode the severity the user would specify.

This was due to the config.get() function getting the defaultObject when there was no validaterc found, this default object specified the same rules as the default validaterc and would take priority. So I added a new flag to prevent this defaultObject from overriding it so it acts as you should expect it to like 3.

dpopp07 reviewed (changes requested):
This looks good! I left a few minor comments but the primary reason I am "requesting changes" is because we need to have some tests for this behavior to ensure it does what we expect it to and to prevent regressions as we keep working on the Spectral stuff

rmkeezer commented on package.json:
```diff
@@ -8,7 +8,8 @@
     "unlink": "npm uninstall -g",
-    "test": "jest test/",
+    "test": "jest test/ && jest test/spectral/tests/spectral-validator.test.js --runInBand",
+    "test:seq": "jest test/spectral/tests/spectral-validator.test.js --runInBand",
```
I added tests to check reading the default config .validaterc, and the user config .spectral.yml, but to test them I had the test create a temporary config file.

However, this interferes with other tests when run in parrallel, so I have this test file running sequentially after the regular tests.

Is there a better solution for tests like these or is it fine to run this file sequentially?

jorge-ibm replied:
Maybe try to spy on the fs package? https://stackoverflow.com/questions/50066138/mock-fs-function-with-jest . Not sure, @dpopp07 might have more experience doing something like that.

dpopp07 commented on test/spectral/tests/spectral-validator.test.js:
```diff
@@ -108,3 +112,155 @@ describe('spectral - test spectral-validator.js', function() {
   });
 });
+
+const sequential = process.argv.includes('--runInBand');
```
My only question on the tests is - is there a way to just look at what rules spectral is ready to use after it's defined? Then we could just call the setup function and make sure it does what we want it to. I think that would be much simpler.

That said, I do like the idea of having an end-to-end test to verify the behavior we're looking for in a real life scenario. We can discuss later

rmkeezer replied:
I was able to spyOn the config files so the end-to-end test isn't as troublesome anymore. For non-end-to-end tests, it's a bit difficult with spectral since creating a ruleset needs a file first and add/modifying rules is a bit messy using spectral's functions.

dpopp07 reviewed (changes requested):
Getting better but I think there are some simplifications to be made that I did not notice before, specifically around reading the config object. Let me know what you think

rmkeezer commented:
I updated the PR with @dpopp07 's suggestions for using the defaultObject instead of calling config.get() inside the spectral run. This worked well! I also updated the unit test to using jest.spyOn to replace the file URL and the returned config file with their respective spectral changes. So the test does not need to run sequentially anymore!

dpopp07 commented on src/spectral/utils/spectral-validator.js:
```diff
@@ -57,12 +58,22 @@ const setup = async function(spectral) {
+  await spectral.loadRuleset([defaultSpectralRulesetURI, spectralRulesetURI]);
+
+  // Combine default/user ruleset with the validaterc spectral rules
+  // The validaterc rules will take precendence in the case of duplicate ruless
```
```suggestion
  // The validaterc rules will take precendence in the case of duplicate rules
```

dpopp07 commented on src/cli-validator/utils/processConfiguration.js:
```diff
@@ -10,6 +10,8 @@ const defaultConfig = require('../../.defaultsForValidator');
 // global objects
 const readFile = util.promisify(fs.readFile);
 const defaultObject = defaultConfig.defaults;
+// Clear spectral rules from defaultObject
```
I think this makes sense but I was confused at first - maybe just add a bit more info in the comment like 
```suggestion
// Clear spectral rules from defaultObject, only use the explicit values in the validaterc file
// the default values for spectral come from another file
```

jorge-ibm reviewed (commented):
Looks really good! My last recommendation would be to add at least 1 test case that uses the ``commandLineValidator`` vs the ``inCodeValidator`` to invoke the validator.

dpopp07 reviewed (approved):
Looks good! I left some extremely minor comments to address before merging but I think it's good otherwise

rmkeezer commented on test/spectral/tests/spectral-validator.test.js:
```diff
@@ -108,3 +114,157 @@ describe('spectral - test spectral-validator.js', function() {
+  // There should be no errors and 2 warnings for a rule not off
+  it('test no errors and no warnings', function() {
+    expect(validationResults.errors.length).toBe(0);
+    expect(validationResults.warnings.length).toBe(2);
```
Yup, there is a warning for `Definition was declared but never used in document`, i'm clarifying the comment for this

mkistler reviewed (approved):
Looks good! 👍

dpopp07 commented:
:tada: This PR is included in version 0.31.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/ibm-openapi-validator/v/0.31.0)
- [GitHub release](https://github.com/IBM/openapi-validator/releases/tag/v0.31.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
